### PR TITLE
[SPARK-30230][SQL] Like ESCAPE syntax can not use '_' and '%'

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Since [25001](https://github.com/apache/spark/pull/25001), spark support like escape syntax. 
But '%' and '_' is the reserve char in `Like` expression. We can not use them as escape char.

### Why are the changes needed?

Avoid some unexpect problem when using like escape syntax.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Add UT.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @beliefer @cloud-fan @maropu @gengliangwang

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -1393,6 +1393,10 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
           }
           str.charAt(0)
         }.getOrElse('\\')
+        if (Set('%', '_').contains(escapeChar)) {
+            throw new ParseException("Invalid escape string." +
+              "Escape string can not be '%', '_'.", ctx)
+        }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: We cannot avoid this case in the antlr rules (SqlBase.g4)?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: no, the antlr side just mach a string literal.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: You mean like this ?
```
ESCAPE_STRING
    : '\'' ( ~('\''|'\\') | ('\\' .) | ~('%' | '_') )* '\''
    | '\'' ( ~('\''|'\\') | ('\\' .) | ~('%' | '_') )* '\''
    ;

NOT? kind=LIKE pattern=valueExpression (ESCAPE escapeChar=ESCAPE_STRING)?
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: seems much easier to do at scala side...

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: with better error message.

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
add to whitelist

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Yes, I do agree.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by ulysses-you** [comment]
Sorry for a mistake.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -1393,6 +1393,10 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
           }
           str.charAt(0)
         }.getOrElse('\\')
+        if (Set('%', '_').contains(escapeChar)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit: don't use `Set` for just 2 elements. We can just do `a = % || a = _`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: ok.

**Review by maropu — APPROVED** [review]
LGTM except for the @cloud-fan comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Done.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by ulysses-you** [comment]
Thanks.

**Comment by beliefer** [comment]
@ulysses-you @maropu @cloud-fan @gengliangwang 
I checked postgreSQL, it supports as below:
`SELECT 'abC' LIKE 'ab%C' ESCAPE '%';`

**Comment by gengliangwang** [comment]
@beliefer you are right. I run the following in PostgreSQL:
```
> SELECT 'abcd' LIKE 'ab%_d';
true
> SELECT 'abcd' LIKE 'ab%_d' escape '%';
false
```

I should not merge it without checking on PostgreSQL. I am going to revert this one. 
Any thoughts? @maropu @cloud-fan

**Comment by cloud-fan** [comment]
Can we check more systems? it does look weird to support % and _ as escape chars.

**Comment by ulysses-you** [comment]
If '%' and '_' as escape, how to use the real pattern.
Like this ?
```
SELECT 'abcd' LIKE 'ab%%' escape '%';
```

But actually `StringUtils.escapeLikeRegex` will do this check.
```
while (in.hasNext) {
      in.next match {
        case c1 if c1 == escapeChar && in.hasNext =>
          val c = in.next
          c match {
            case '_' | '%' => out ++= Pattern.quote(Character.toString(c))
            case c if c == escapeChar => out ++= Pattern.quote(Character.toString(c))
            case _ => fail(s"the escape character is not allowed to precede '$c'")
          }
        case c if c == escapeChar => fail("it is not allowed to end with the escape character")
        // we will never run these code path
        case '_' => out ++= "."
        case '%' => out ++= ".*"
        case c => out ++= Pattern.quote(Character.toString(c))
      }
    }
```

**Comment by beliefer** [comment]
```
>SELECT 'abcd' LIKE 'ab%%';
true
>SELECT 'abcd' LIKE 'ab%%' ESCAPE '%';
false
```

**Comment by cloud-fan** [comment]
@gengliangwang let's revert this.

**Comment by maropu** [comment]
+1 for the revert, thanks for taking care, @gengliangwang !

**Comment by ulysses-you** [comment]
OK,  I see.  Thanks for reverting.

**Pull request closed without merging**